SimPEG/base/pde_simulation.py

```diff
@@ -1,41 +1,72 @@
 import numpy as np
 import scipy.sparse as sp
-from discretize.utils import Zero
+from discretize.utils import Zero, TensorType
 from ..simulation import BaseSimulation
 from .. import props
 from scipy.constants import mu_0
 
 
 def __inner_mat_mul_op(M, u, v=None, adjoint=False):
     u = np.squeeze(u)
-    if v is not None:
-        if v.ndim > 1:
-            v = np.squeeze(v)
-        if u.ndim > 1:
-            # u has multiple fields
-            if v.ndim == 1:
-                v = v[:, None]
-        else:
+    if sp.issparse(M):
+        if v is not None:
             if v.ndim > 1:
-                u = u[:, None]
-        if v.ndim > 2:
-            u = u[:, None, :]
-        if adjoint:
-            if u.ndim > 1 and u.shape[-1] > 1:
-                return M.T * (u * v).sum(axis=-1)
-            return M.T * (u * v)
-        if u.ndim > 1 and u.shape[1] > 1:
-            return np.squeeze(u[:, None, :] * (M * v)[:, :, None])
-        return u * (M * v)
+                v = np.squeeze(v)
+            if u.ndim > 1:
+                # u has multiple fields
+                if v.ndim == 1:
+                    v = v[:, None]
+            else:
+                if v.ndim > 1:
+                    u = u[:, None]
+            if v.ndim > 2:
+                u = u[:, None, :]
+            if adjoint:
+                if u.ndim > 1 and u.shape[-1] > 1:
+                    return M.T * (u * v).sum(axis=-1)
+                return M.T * (u * v)
+            if u.ndim > 1 and u.shape[1] > 1:
+                return np.squeeze(u[:, None, :] * (M * v)[:, :, None])
+            return u * (M * v)
+        else:
+            if u.ndim > 1:
+                UM = sp.vstack([sp.diags(u[:, i]) @ M for i in range(u.shape[1])])
+            else:
+                U = sp.diags(u, format="csr")
+                UM = U @ M
+            if adjoint:
+                return UM.T
+            return UM
     else:
+        # assume it was a tuple of M_func, prop_deriv
+        M_func, prop_deriv = M
         if u.ndim > 1:
-            UM = sp.vstack([sp.diags(u[:, i]) @ M for i in range(u.shape[1])])
+            Mu = [M_func(u[:, i]) for i in range(u.shape[1])]
+            if v is None:
+                Mu = sp.vstack([M @ prop_deriv for M in Mu])
+                if adjoint:
+                    Mu = Mu.T
+                return Mu
+            elif v.ndim > 1:
+                v = np.squeeze(v)
+            if adjoint:
+                return sum(
+                    [prop_deriv.T @ (Mu[i].T @ v[..., i]) for i in range(u.shape[1])]
+                )
+            pv = prop_deriv @ v
+            return np.stack([M @ pv for M in Mu], axis=-1)
         else:
-            U = sp.diags(u, format="csr")
-            UM = U @ M
-        if adjoint:
-            return UM.T
-        return UM
+            Mu = M_func(u)
+            if v is None:
+                Mu = Mu @ prop_deriv
+                if adjoint:
+                    Mu = Mu.T
```

Comment:
same comment as above about returning rather than re-assigning
```suggestion
                    return Mu.T
```